plugins/techdocs-backend/src/cache/TechDocsCache.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { CacheClient } from '@backstage/backend-common';
+import { assertError, CustomErrorBase } from '@backstage/errors';
+import { Config } from '@backstage/config';
+import { Logger } from 'winston';
+
+export class CacheInvalidationError extends CustomErrorBase {}
+
+export class TechDocsCache {
+  protected readonly cache: CacheClient;
+  protected readonly logger: Logger;
+  protected readonly readTimeout: number;
+
+  private constructor({
+    cache,
+    logger,
+    readTimeout,
+  }: {
+    cache: CacheClient;
+    logger: Logger;
+    readTimeout: number;
+  }) {
+    this.cache = cache;
+    this.logger = logger;
+    this.readTimeout = readTimeout;
+  }
+
+  static fromConfig(
+    config: Config,
+    { cache, logger }: { cache: CacheClient; logger: Logger },
+  ) {
+    const readTimeout =
+      config.getOptionalNumber('techdocs.cache.readTimeout') || 1000;
+    return new TechDocsCache({ cache, logger, readTimeout });
+  }
+
+  async get(path: string): Promise<Buffer | undefined> {
+    try {
+      // Promise.race ensures we don't hang the client for long if the cache is
+      // temporarily unreachable.
+      const response = (await Promise.race([
+        this.cache.get(path),
+        new Promise(cancelAfter => setTimeout(cancelAfter, this.readTimeout)),
+      ])) as string | undefined;
+
+      if (response !== undefined) {
```

Comment:
By conflating the origin of `undefined` coming from either the `setTimeout` or an empty `cache.get`, I think we might miss catching errors in the underlying cache. It would be worthwhile to have the option to monitor if 100% of cache misses are caused by timeouts or return immediately and hint to errors with the cache itself

Reply:
I'd love to get this level of detail in some kind of observability framework, since even as is, it's only `debug`-level log statements, which probably won't be captured anywhere in production.   For now, you can get a sense of when things go awry when response times start to be suspiciously correlated to the `readTimeout` configuration value (I've seen blips of this in our internal deployment of this patch-set, for example).